### PR TITLE
Example: Linear Elastic Global pROM

### PR DESCRIPTION
**Pull request by axla-io**
# Example: Linear Elastic Global pROM

## Background and description
As an exercise in getting to know the `libROM `library, I implemented a global pROM for a linear elastic cantilever problem.

The program is based on [ex2p](https://github.com/mfem/mfem/blob/master/examples/ex2p.cpp) in MFEM and the `poisson_global_rom` example in `libROM`.

## What's added

The files:
* `examples/data/beam-tri.mesh` - where the beam mesh from [ex2p](https://github.com/mfem/mfem/blob/master/examples/ex2p.cpp) is defined.
* `examples/prom/linear_elasticity_global_rom.cpp` - which is the main program for the pROM example

The program has support for both external force parametrization and Young's Modulus $E$ + Poisson's ratio $\nu$ parametrization.

## How has it been tested?
The code runs without any problems for the command line options offline, online, `merge`, `fom`, `visit `+ the added parameterization options.

I have made relative error comparisons for the predicted ROM solutions with the FOM solutions and they are consistent with expected results.

## Comments
I have tried to keep my code consistent with the libROM syntax. Feel free to make or suggest changes.
Some unnecessary line breaks were added to some files. I saw that just now. Hope that it isn't too much of a nuisance.

## Reviewers
@chldkdtn @dylan-copeland @kevinhkhuynh @siuwuncheung

### CODE REVIEW

**Comment by chldkdtn** [comment]
This is Axel's first contribution to libROM. Kudos to @axla-io 

@dylan-copeland @kevinhkhuynh @siuwuncheung let's review and see if there can be any improvements we can make here. This is a good first example for linear elasticity as well!

File: examples/prom/linear_elasticity_global_rom.cpp
Comment (dylan-copeland) [inline examples/prom/linear_elasticity_global_rom.cpp:7]: This needs to be edited.

File: examples/prom/linear_elasticity_global_rom.cpp
Comment (dylan-copeland) [inline examples/prom/linear_elasticity_global_rom.cpp:127]: Indenting is not consistent. In libROM, we run `./scripts/stylize.sh` with astyle 2.05 (or 3.1, we should document this in the README) to ensure consistent formatting. Let us know if you need help with astyle.

File: examples/prom/linear_elasticity_global_rom.cpp
Comment (dylan-copeland) [inline examples/prom/linear_elasticity_global_rom.cpp:141]: I'm not sure if astyle will remove these extra empty lines. In any case, our convention, which follows the MFEM conventions, is not to have more than 1 empty line.

File: examples/prom/linear_elasticity_global_rom.cpp
Comment (dylan-copeland) [inline examples/prom/linear_elasticity_global_rom.cpp:520]: This output can be removed.

File: cmake/toolchains/default-toss_3_x86_64_ib-librom-dev.cmake
Comment (dylan-copeland) [inline cmake/toolchains/default-toss_3_x86_64_ib-librom-dev.cmake:27]: This change is not necessary.

**Comment by dylan-copeland** [comment]
We usually work with branches in libROM rather than forks like this PR. It doesn't matter very much, but one inconvenience of the fork is that reviewers have to clone the fork instead of just checking out a branch. Please feel free to use branches in the future.

File: examples/dmd/nonlinear_elasticity.cpp
Comment (dylan-copeland) [inline examples/dmd/nonlinear_elasticity.cpp:77]: This change is not necessary.

File: cmake/toolchains/default-toss_3_x86_64_ib-librom-dev.cmake
Comment (axla-io) [reply cmake/toolchains/default-toss_3_x86_64_ib-librom-dev.cmake:27]: Thanks! I'll remove the extra empty lines. Sorry about these ones.

File: examples/dmd/nonlinear_elasticity.cpp
Comment (axla-io) [reply examples/dmd/nonlinear_elasticity.cpp:77]: Same here.

File: examples/prom/linear_elasticity_global_rom.cpp
Comment (axla-io) [reply examples/prom/linear_elasticity_global_rom.cpp:7]: I'll fix it!

File: examples/prom/linear_elasticity_global_rom.cpp
Comment (axla-io) [reply examples/prom/linear_elasticity_global_rom.cpp:127]: Thanks! I hadn't seen that before. I'll try it out and let you know.

File: examples/prom/linear_elasticity_global_rom.cpp
Comment (axla-io) [reply examples/prom/linear_elasticity_global_rom.cpp:141]: Thanks, I'll keep that in mind.

File: examples/prom/linear_elasticity_global_rom.cpp
Comment (axla-io) [reply examples/prom/linear_elasticity_global_rom.cpp:520]: Thanks for the catch!

File: examples/prom/linear_elasticity_global_rom.cpp
Comment (axla-io) [reply examples/prom/linear_elasticity_global_rom.cpp:7]: Just a question, what is it in particular that I should change about the description?

**Comment by axla-io** [comment]
> We usually work with branches in libROM rather than forks like this PR. It doesn't matter very much, but one inconvenience of the fork is that reviewers have to clone the fork instead of just checking out a branch. Please feel free to use branches in the future.

Thanks for the input. I'll keep this in mind for the future.

File: examples/prom/linear_elasticity_global_rom.cpp
Comment (axla-io) [reply examples/prom/linear_elasticity_global_rom.cpp:126]: In the latest commit I used `astyle `to format the code.

**Comment by chldkdtn** [comment]
> > We usually work with branches in libROM rather than forks like this PR. It doesn't matter very much, but one inconvenience of the fork is that reviewers have to clone the fork instead of just checking out a branch. Please feel free to use branches in the future.
> 
> Thanks for the input. I'll keep this in mind for the future.

@axla-io I see that you have not accepted the invitation to the access or collaborator request to this libROM repo development. Please make sure you accept the invitation before it get expired.

**Comment by axla-io** [comment]
> @axla-io I see that you have not accepted the invitation to the access or collaborator request to this libROM repo development. Please make sure you accept the invitation before it get expired.

@chldkdtn Thanks, I have accepted now.

File: examples/prom/linear_elasticity_global_rom.cpp
Comment (dylan-copeland) [reply examples/prom/linear_elasticity_global_rom.cpp]: The description is fine, it just needs the indenting aligned.

File: examples/prom/linear_elasticity_global_rom.cpp
Comment (dylan-copeland) [inline examples/prom/linear_elasticity_global_rom.cpp:402]: In the online phase, or in a post-processing phase, it would be nice to compute the error of the ROM solution (lifted to FOM space) with respect to a FOM solution stored in a file. That is more complicated and I suppose not necessary for this PR, but usually that is desired to verify the ROM is accurate.

**Review by dylan-copeland — APPROVED** [review]

File: examples/prom/linear_elasticity_global_rom.cpp
Comment (axla-io) [reply examples/prom/linear_elasticity_global_rom.cpp:27]: Ooops! A typo. I'll get that fixed!

File: examples/prom/linear_elasticity_global_rom.cpp
Comment (axla-io) [reply examples/prom/linear_elasticity_global_rom.cpp:496]: Thanks! I'll remove them.

File: examples/prom/linear_elasticity_global_rom.cpp
Comment (axla-io) [reply examples/prom/linear_elasticity_global_rom.cpp:402]: @dylan-copeland @chldkdtn I agree that it'd be nice to have and I already did the function to calculate the relative error, but in a standalone script. I can adapt it to be used in the online phase instead.

Comment (kevinhkhuynh) [reply examples/prom/linear_elasticity_global_rom.cpp:402]: If it helps, in lines 889-901 of this other example, the relative error is also computed: https://github.com/LLNL/libROM/blob/master/examples/prom/dg_advection_global_rom.cpp

File: examples/prom/linear_elasticity_global_rom.cpp
Comment (siuwuncheung) [reply examples/prom/linear_elasticity_global_rom.cpp:496]: Thanks, Axel! Everything else looks good to me. I am ready to approve the PR once the minor issues are fixed.

File: examples/prom/linear_elasticity_global_rom.cpp
Comment (axla-io) [reply examples/prom/linear_elasticity_global_rom.cpp:405]: Hi! I have now implemented the relative error calculation. Feel free to check it out @dylan-copeland @chldkdtn @kevinhkhuynh

**Review by siuwuncheung — APPROVED** [review]

**Comment by chldkdtn** [comment]
@axla-io You got two approvals, which are sufficient to "squash and merge" You are welcome to do so.

**Comment by axla-io** [comment]
> @axla-io You got two approvals, which are sufficient to "squash and merge" You are welcome to do so.

Thanks! I'll do so then.

**Comment by YuDe95** [comment]
@axla-io I have been confused for a long time and don't know how to predict multiple parameters, such as E and nu in the material. Some examples I have seen use dmd_u ->predict (ts [0]); Can you provide a more concise example on how to use ROM to predict multiple parameters, given the predicted time

**Comment by axla-io** [comment]
Hi @YuDe95, this PR does not provide the capability to do parameter estimation. You seem to be referring to DMD functionality, whereas this PR is for pROM. Also, in the future, I recommended to ask questions like this in an issue, since it's easier to manage for the maintainers and for other users with the same questions to find answers.

**Pull request merged**
